DRA: call plugins for claims even if exist in cache

adrianchiris opened the pull request:
Initial PR to gather Feedback if the proposed approach for fixing #119850 is acceptable or a different approach is required.
See [Alternatives](#alternatives) section for a different approach.

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Today, DRA manager does not call plugin NodePrepareResource
for claims that it previously successfully handled, that is,
if claims are present in cache (checkpoint) even if node
rebooted.

After node reboots, it is required to call DRA plugin
for resource claims so that plugins may prepare them
again in case the resources dont persist reboot.

To achieve that, once kubelet is started, we call DRA
plugins for claims once if a pod sandbox is required
to be created during PodSync.

##### DRA Manager NodePrepareResource  in different scenarios:

When kubelet restarts, DRA Manager NodePrepareResource will not be called during SyncPod as pods will already be in running state (pod sandbox state ready), thus no redundant calls will be performed to DRA plugins.

When A Pod's sandbox need to be (re)created, DRA Manager NodePrepareResource will be called however its claims in in-memory claim cache would be marked as "prepared", thus no call to plugin will occur.

When Kubelet restarts AND a Pod's sandbox need to be (re)created, DRA manager will call DRA plugin. At that point its expected that the plugin is idempotent and return same results as resource already exists.

When Node reboots, resource may not be created on the node. however same steps as above will occur, which will trigger resource preparation by DRA plugin.
Note that result may be different in this case. thus latest results for a claim are stored in checkpoint. 

#### Which issue(s) this PR fixes:

Fixes #119850 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
NONE

#### Alternatives

Place DRA manager checkpoint file under /var/run/kubelet so that its content is cleared
on node boot. that way if a pod sandbox is required to be created (during PodSync), NodePrepareResouce call will only
call DRA plugins for a given resource claim once.

adrianchiris committed:
DRA: call plugins for claims even if exist in cache

Today, DRA manager does not call plugin NodePrepareResource
for claims that it previously successfully handled, that is,
if claims are present in cache (checkpoint) even if node
rebooted.

After node reboots, it is required to call DRA plugin
for resource claims so that plugins may prepare them
again in case the resources dont persist reboot.

To achieve that, once kubelet is started, we call DRA
plugins for claims once if a pod sandbox is required
to be created during PodSync.

Signed-off-by: adrianc <adrianc@nvidia.com>

adrianchiris commented:
/sig node

adrianchiris commented:
cc @moshe010 @klueska

bart0sh commented:
/triage accepted
/priority important-soon

adrianchiris commented:
@bart0sh @klueska PTAL and let me know if you prefer the alternative.

bart0sh commented:
I generally like this approach.
It would be good though, to understand if we need to put other checkpoints (device plugins, cpu/memory/topology managers) to /var/run/kubelet.
Another thing that this change makes `rundir` a command like option, which means that users can effectively disable checkpoint removal without knowing about it.  We should probably mention this in the command line option description or make `rundir` a constant.

adrianchiris commented:
> It would be good though, to understand if we need to put other checkpoints (device plugins, cpu/memory/topology managers) to /var/run/kubelet.

perhaps @klueska has some inputs on this one ? (will try to have a look , im not too familiar with the other managers)
also i think each such manager would need to be addressed in his own PR. 

> Another thing is that this change makes rundir a command like option, which means that users can effectively disable checkpoint removal without knowing about it. We should probably mention this in the command line option description or make rundir a constant.

im fine with either, for now, let me update flag description that it is a requirement this path is cleared on node boot by some means.

bart0sh commented:
/lgtm

bart0sh commented:
/assign @klueska

klueska commented:
Sorry it's taken me so long to get to review this.

I'm not sure I agree with the following premise which motivates this PR:
> After node reboots, it is required to call DRA plugin for resource claims so that plugins may prepare them again in case the resources don't persist reboot.

It should be up to the plugin to maintain its own state about what it has prepared or not (how else could it be idempotent), and if that state is not the current state of the hardware when the plugin tries to come online (for example after a node reboot or a simple restart), then it needs to do what's necessary to bring the hardware to that state before it marks itself as ready.

adrianchiris commented:
> It should be up to the plugin to maintain its own state about what it has prepared or not (how else could it be idempotent),

i assume it depends on the plugin, one way is to have the desired result in the request. (e.g in my case request consists of allocating a scalable function aka SF on a specific device with a specific index)

> then it needs to do what's necessary to bring the hardware to that state before it marks itself as ready.

even if a plugin stores it state in yet another cache. how do you ensure pods consuming resources dont spin up (by kubelet on node reboot) before the dra plugin pod starts and has time to align its own internal state.

now, even if it has time to sort its internal state by some means, what if preparing some of the resources fail (and some succeed) do we fail all by not starting ?

there are even more subtleties in my case when we get into creating resources in parallel but i think these are solvable

adrianchiris commented:
OK, I have discussed with @klueska off-line, the points raised above are valid in his view.

However his preference is to take the alternative approach described in the PR title.
as its a more conservative approach, which seems safer as currently we dont have anything kubelet related in /var/run.

As kubelet should generally NOT restart, the added overhead of calling DRA plugins on kubelet restart is the same as node reboot and should be deemed acceptable IMO.

@bart0sh LMK if you are also OK with this (the alternative) approach and I can start working on implementation.

bart0sh commented:
> @bart0sh LMK if you are also OK with this (the alternative)

I'm generally OK, but would like to know more details about possible overhead, i.e. when and how often `NodePrepareResources` will be called in this case in comparison with the original approach.

Would it make sense to implement a node_e2e to simulate node reboot and analyse both approaches?

adrianchiris committed:
Add unit tests

adjust existing tests and add new test flows
to cover new DRA manager behaviour

Signed-off-by: adrianc <adrianc@nvidia.com>

adrianchiris commented:
>I'm generally OK, but would like to know more details about possible overhead, i.e. when and how often NodePrepareResources will be called in this case in comparison with the original approach.

Hi, appologies for the delay, i have updated the PR to reflect the new direction. see updated PR description as well.

**The below compares the two approaches:**

**Approach-1(Original)**: Store checkpoint in /var/run/
**Approach-2 (Current)**: Call dra plugin for pod claims regardless of their presence in cache, but keep track in-memory if kubelet already prepared a given claim so we dont process a claim twice during kubelet run.

given a single pod running on node with N claims to K plugins.

**Scenarios:**
1. Pod is already running, kubelet restarted

both approach 1 and 2 will not call DRA plugins for pod claims.

2. pod sandbox need to be re-created, kubelet running

both approach 1 and 2 will not call DRA plugins for pod claims.

3. kubelet restarted then Pod sandbox need to be re-created

Approach-1 will not call DRA plugins for pod claims
Approach-2 will call DRA plugins for pod claims (K calls to DRA plugins), subsequent pods that reference same claims will not trigger a call to DRA pluigns.

4. node reboot

Both Approach-1 and Approach-2 will call DRA plugins  for pod claims.
subsequent pods that reference same claims will not trigger a call to DRA pluigns


so essentially the only overhead WRT DRA plugins is in scenario 3, as kubelet restart would cause kubelet to "lose" its in-memory state of which claims it handled. it may require up to N claims to be prepared by DRA plugins, N being the total number of claims that are bound to this node. IF we start having pod sandbox recreation requests for Pods.

adrianchiris commented:
/retest

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -120,31 +122,34 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-		if len(resourceHandles) == 0 {
-			resourceHandles = make([]resourcev1alpha2.ResourceHandle, 1)
-		}
+			if claimInfo.preparedByManager {
+				continue
+			}
+		} else {
+			// Grab the allocation.resourceHandles. If there are no
+			// allocation.resourceHandles, create a single resourceHandle with no
+			// content. This will trigger processing of this claim by a single
+			// kubelet plugin whose name matches resourceClaim.Status.DriverName.
+			resourceHandles := resourceClaim.Status.Allocation.ResourceHandles
+			if len(resourceHandles) == 0 {
+				resourceHandles = make([]resourcev1alpha2.ResourceHandle, 1)
+			}
```
Something feel odd about this -- it may be correct, but the logic is not very obivous and I'm getting a little lost trying to understand it.

Previously the logic was:
* If the claim is in the cache, we know it has been prepared successfully, so just add a pod reference and continue
* If the claim is not in the cache, attempt to prepare it, and if that succeeds, add it to the cache.

I think the new logic should be:
* If the claim is in the cache and it is prepared, add a pod reference and continue
* If the claim is in the cache, and not prepared, prepare it and mark it as such
* If the claim is not in the cache, prepare it, add it to the cache, and make it as prepared

If you agree, can you update the logic to make this flow more obvious.

adrianchiris replied:
i agree at high level, however the lower level details are a bit different.

one thing to note is that checking if a claim needs to be processed  (prepared) by the manager and marking it as prepared then actually adding to cache happens in a separate loops

the method has two main loops:
1. go over all pod claims, check them and determine which claims should process by calling to DRA plugin  (filling in `claimInfos` map and `batches`)
2. call DRA plugins in a batch manner and update cache

the changes i introduced:
in 1 :  claims that have not been processed (not prepared) by manager will get added to the relevant structs to be processed
in 2: after a claim has been successfully handled by DRA plugin mark it as prepared before adding it to cache.

lemme try to simplify 1 a bit.

adrianchiris reviewed (commented):
Thx for the feedback, will upload PR shortly

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -109,42 +107,31 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-		// content. This will trigger processing of this claim by a single
-		// kubelet plugin whose name matches resourceClaim.Status.DriverName.
-		resourceHandles := resourceClaim.Status.Allocation.ResourceHandles
-		if len(resourceHandles) == 0 {
-			resourceHandles = make([]resourcev1alpha2.ResourceHandle, 1)
+		if claimInfo == nil {
```
nit: remove newline

adrianchiris replied:
Done.

klueska commented on pkg/kubelet/cm/dra/manager.go:
```diff
@@ -109,42 +107,31 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
-		if len(resourceHandles) == 0 {
-			resourceHandles = make([]resourcev1alpha2.ResourceHandle, 1)
+		if claimInfo == nil {
+			// claim does not exist in cache, create new claimInfo object
+			// to be processed later
```
```suggestion
			// Claim does not exist in cache, create new claimInfo object
			// to be processed later.
```

adrianchiris replied:
done.

klueska reviewed (commented):
Thanks for the cleanup. It is much clearer now what is going on.

/approve
/lgtm

klueska commented:
/lgtm

pacoxu commented:
A release note is needed.
The original release note is like
```
Added a new kubelet CLI flag "--run-dir" : directory path for managing kubelet runtime files. (default "/var/run/kubelet")
```

Or you may comment `/release-note-none` if no release note is not needed.

adrianchiris commented:
no release note is needed for this one (went with a different impl than the original proposal)

/release-note-none

adrianchiris commented:
@bart0sh @klueska do we want to backport this to 1.28 ? is there anything i need to do WRT backport ?
